projects/driver_controls/test/test_can_output.c

```diff
@@ -0,0 +1,299 @@
+#include "turn_signal_fsm.h"
+
+#include "can_output.h"
+
+typedef struct FSMGroup {
+  FSM power;
+  FSM pedal;
+  FSM direction;
+  FSM turn_signal;
+  FSM hazard_light;
+  FSM mechanical_brake;
+  FSM horn;
+  FSM push_to_talk;
+} FSMGroup;
+
+static FSMGroup s_fsm_group;
+static bool s_powered = false;
+static bool s_mech_brake = false;
+
+static EventArbiterOutputData s_can_output;
+
+static void prv_output(EventArbiterOutputData data) {
+  s_can_output.id = data.id;
+  s_can_output.state = data.state;
+  s_can_output.data = data.data;
+
+  LOG_DEBUG("[ .id = %x, .state = %x, .data = %x ]\n",
+          s_can_output.id,
+          s_can_output.state,
+          s_can_output.data);
+}
+
+static void prv_toggle_power(bool new_state) {
+  Event e = { .data = 0 };
+  if (new_state != s_powered) {
+    e.id = INPUT_EVENT_POWER;
+    event_arbiter_process_event(&e);
+    s_powered = new_state;
+  }
+}
+
+static void prv_toggle_mech_brake(bool new_state) {
+  Event e = { .data = 0 };
+
+  e.id = (new_state == true) ? INPUT_EVENT_MECHANICAL_BRAKE_PRESSED
+                             : INPUT_EVENT_MECHANICAL_BRAKE_RELEASED;
+
+  event_arbiter_process_event(&e);
+  s_mech_brake = new_state;
+}
+
+void setup_test(void) {
+  event_arbiter_init(prv_output);
+
+  power_fsm_init(&s_fsm_group.power);
+  pedal_fsm_init(&s_fsm_group.pedal);
+  direction_fsm_init(&s_fsm_group.direction);
+  turn_signal_fsm_init(&s_fsm_group.turn_signal);
+  hazard_light_fsm_init(&s_fsm_group.hazard_light);
+  mechanical_brake_fsm_init(&s_fsm_group.mechanical_brake);
+  horn_fsm_init(&s_fsm_group.horn);
+  push_to_talk_fsm_init(&s_fsm_group.push_to_talk);
+
+  s_powered = false;
+  s_mech_brake = false;
+
+  event_queue_init();
+}
+
+void teardown_test(void) { }
+
+// Test that the power fsm correctly generates CAN events
+void test_can_output_power(void) {
+  prv_toggle_power(true);
+
+  TEST_ASSERT_EQUAL(CAN_OUTPUT_MESSAGE_POWER, s_can_output.id);
+  TEST_ASSERT_EQUAL(POWER_FSM_STATE_ON, s_can_output.state);
+  TEST_ASSERT_EQUAL(0, s_can_output.data);
+
+  prv_toggle_power(false);
+
+  TEST_ASSERT_EQUAL(CAN_OUTPUT_MESSAGE_POWER, s_can_output.id);
+  TEST_ASSERT_EQUAL(POWER_FSM_STATE_OFF, s_can_output.state);
+  TEST_ASSERT_EQUAL(0, s_can_output.data);
+}
+
+void test_can_output_mechanical_brake(void) {
+  // Test that pressing the brake state generates the correct event
+  prv_toggle_mech_brake(true);
+
+  TEST_ASSERT_EQUAL(CAN_OUTPUT_MESSAGE_MECHANICAL_BRAKE, s_can_output.id);
+  TEST_ASSERT_EQUAL(MECHANICAL_BRAKE_FSM_STATE_ENGAGED, s_can_output.state);
+  TEST_ASSERT_EQUAL(0, s_can_output.data);
+
+  // Test that releasing the brake generates the correct event
+  prv_toggle_mech_brake(false);
+
+  TEST_ASSERT_EQUAL(CAN_OUTPUT_MESSAGE_MECHANICAL_BRAKE, s_can_output.id);
+  TEST_ASSERT_EQUAL(MECHANICAL_BRAKE_FSM_STATE_DISENGAGED, s_can_output.state);
+  TEST_ASSERT_EQUAL(0, s_can_output.data);
+}
+
+void test_can_output_hazard_light(void) {
+  Event e = { 0 };
```

Comment:
NIT: Since this is used so widely you may as well define it as static global and just reset it in the init.

Reply:
Personal opinion: Not a fan of static variables. I'd like to avoid them in our code where we can.